---
title: "CellRomeR short introduction"
authors: ""
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{CellRomeR short introduction}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding[utf8]{inputenc}
---
  
<style>
body {
text-align: justify}
</style>
  
```{r knitr-options, echo=FALSE, message=FALSE, warning=FALSE}
library(knitr)
opts_chunk$set(fig.align = 'center', fig.width = 7, fig.height = 5)
```

# Introduction 

CellRomeR is an R package for the analysis and visualisation of cell tracking data from TrackMate [1]. It uses ILoReg [2] for clustering and includes several different plotting functions. Input data is an XML file from TrackMate.

# Installation

CellRomeR is available on Github. It can be installed using devtools and the following commands

```{r installation, eval=FALSE, include=FALSE}
#install.packages("devtools")
#devtools:install_github("elolab/cellromer")
```

# Example analysis

```{r libraries, include=FALSE}
library(CellRomeR)
library(data.table)
library(ILoReg)
```

## Data Import

XML is the native data format used by TrackMate to store data and the models. In TrackMate data, long tracks can often branch one or more times due to cell divisions or tracking issues. Plotting these branching tracks can pose challenges. To address this, it is advisable to only export non-branching tracks from TrackMate. We provide the `IRmigr.TMxml()` function to read single TrackMate data into `MigrDat` migration data object. It reads and stores spots, tracks, and edges data in separate slots in a `MigrDat` object.

An example XML output file form TrackMate is available here: LINK
The example data is based on the cancer cell migration dataset available in [Zenodo](https://zenodo.org/records/5206107). The original trackMate `P31-crop.xml` dataset was imported into TrackMate (load in TrackMate file), re-tracked, and saved with as non-branching version (`P31-crop-ik.xml`).


```{r TMxml}
#migrdata <- import_XML("P31-crop_ik.xml")
#migrdata <- import_XML("~/Spatial_Migration/Ext_data/1_Random_migration_HeLa_cells.xml")
#migrdata <- import_XML("~/Spatial_Migration/Ext_data/1_Random_migration_HeLa_cells_frames0_25.xml")
migrdata <- import_XML("~/Spatial_Migration/Ext_data/1_Random_migration_HeLa_Cellpose_LapT_Fltred.xml")
migrdata 

```

```{r}
migrdata <- readRDS("~/Spatial_Migration/Ext_data/1_Random_migration_HeLa_Cellpose_LapT_Clust_All.Rds")
migrdata 

migrdata@clustering

CellRomeR:::default.ILoRegclusters(migrdata) 
CellRomeR:::default.PCAs(migrdata)[[1]] 

names(migrdata@dimreductions$S)
names(migrdata@clustering$S)

grep("I[Ll]oReg", colnames(migrdata@clustering$S),value = T) 

CellRomeR:::default.ILoRegclusters(migrdata)[[1]]

```


## Filtering 

Unfortunately is still needed. 

- TrackMate extra spots 
- Short tracks 


```{r filtering}

#CellRomeR::tm_track_filter(migrdata) 
#migrdata@metadata$track_filter

#migrdata <- CellRomeR::rm_unmated_spots(migrdata) 

#tm_filter <- migrdata@metadata$track_filter 
#migrdata <- CellRomeR::filter.tracks(migrdata, track.slot = "raw", filter = TRACK_ID %in% tm_filter) 

#migrdata <- CellRomeR::filter.spots(migrdata, filter = TRACK_ID != "ninTracks", spot.slot = "raw") 
#CellRomeR::tm_track_filter(migrdata)
plot(migrdata)
#print(CellRomeR::spots.raw(migrdata))
#table(CellRomeR::spots.raw(migrdata)[["TRACK_ID"]])

migrdata <- Apply_TMate_Filter(migrdata) 

plot(migrdata)



```


## Clustering with selected features  

```{r varselection}

# CellRomeR::MigrObj.vars.raw(migrdata)$spot.vars
# clustering.vars <- CellRomeR::MigrObj.vars.raw(migrdata)$spot.vars[c(10,12,13,22:31,33)] 

clustering.vars <- colnames(migrdata@spots$raw)[c(10,12,13,22:31,33)] 

head(migrdata@spots$raw)

migrdata <- clustering(migrdata, dat.slot = "raw", type = "S", uniq = "by_most", kILoReg = 5, vars = clustering.vars) 
plot_umap(migrdata, uniq="by_most")

saveRDS(migrdata, "~/Spatial_Migration/Ext_data/1_Random_migration_HeLa_Cellpose_LapT_ClustWMost.Rds")

#clustering.vars <- CellRomeR::MigrObj.vars.raw(migrdata)$spot.vars[c(10,12:17,20:31,33)] 
clustering.vars <- colnames(migrdata@spots$raw)[c(10,12:17,20:31,33)] 

migrdata <- clustering(migrdata, dat.slot = "raw", type = "S", uniq = "by_all", kILoReg = 8, vars = clustering.vars) 
plot_umap(migrdata, uniq="by_all")

saveRDS(migrdata, "~/Spatial_Migration/Ext_data/1_Random_migration_HeLa_Cellpose_LapT_Clust_All.Rds")


```


## Clustering with predefined features  

```{r clustering}

# CellRomeR::Std.pattern(migrdata)
#CellRomeR::default.DRs(migrdata)

migrdata <- clustering(migrdata, dat.slot = "raw", type = "S", uniq = "by_morph", kILoReg = 5, predef = "morphological") 

#plot_tsne(migrdata, uniq="by_morph") 
plot_umap(migrdata, uniq="by_morph") 

saveRDS(migrdata, "~/Spatial_Migration/Ext_data/1_Random_migration_HeLa_Cellpose_LapT_morpClust.Rds")

migrdata <- clustering(migrdata, dat.slot = "raw", type = "S", uniq = "by_tech", kILoReg = 5, predef = "technical") 
plot_umap(migrdata, uniq="by_tech") 

saveRDS(migrdata, "~/Spatial_Migration/Ext_data/1_Random_migration_HeLa_Cellpose_LapT_techClust.Rds")


migrdata <- clustering(migrdata, dat.slot = "raw", type = "S", uniq = "CH1", kILoReg = 8, incl.pattern = "CH1", predef = "none") 

plot_umap(migrdata, uniq="CH1") 

saveRDS(migrdata, "~/Spatial_Migration/Ext_data/1_Random_migration_HeLa_Cellpose_LapT_CH1_Clust.Rds")


migrdata@clustering

```


```{r}


plot(migrdata)
plot_rag(migrdata)
plot_heatmap(migrdata)
plot_violin(migrdata, feature="QUALITY")

plot_umap(migrdata, uniq="by_all") # clustering from CH1 
plot_umap(migrdata, uniq="by_most") # clustering from CH1 
plot_umap(migrdata, uniq="by_morph") # clustering from CH1 
plot_umap(migrdata, uniq="CH1") # clustering from CH1 
plot_umap(migrdata, uniq="by_tech") # clustering from CH1



```


## `MigrDat` data object

The data is stored in a MigrDat object which has the following slots:

- Spots      : Containing information related to each spot
- Tracks     : Information regarding tracks and movement of the cells
- Edges      : Information regarding edges ## Iivari please check!
- Roi_points : Defining *outlines of the spots/cells* (regions of interest)
- Metadata   : Containing any extra information that might be useful
- IDs        : The provided IDs from the TrackMate output
- Dimensional 
  reductions : Containing calculated distance coordinates of spots, tracks, and edges
    
## Clustering

The function `clustering()` can be used to run dimensionality reduction and clustering. The clustering is based on ILoReg [2], an ensembl clustering method developed for clustering of single-cell RNA-sequencing data. The argument `uniq` is used to store the clustering result with a unique identifier for later identification. NOTE! Clustering may take several minutes depending on the size of the data and computational resources. 

```{r Clustering} 

#migrdata <- clustering(MigrObj=migrdata, dat.slot="raw", type="S", incl.pattern=c("CH1"), uniq="CH1_15", kILoReg=15)
load("migrdata3.RData")

```

The clustering above was done for Spot (S) type only, but by adjusting the `type` parameter we can also do this for Tracks (T) and Edges (E).

## Visualisation

cellRomeR includes several plotting options. The `plot` function plots the trajectories of the cells. The `plot_violin` can be used on any of the data features. 


```{r clusts}

plot(migrdata)
plot_heatmap(migrdata)
plot_violin(migrdata, feature="QUALITY")
plot_tsne(migrdata, uniq="CH1_15")


```

## Session Info

```{r session}

sessionInfo()
```

# References

1. Ershov, D., Phan, M.-S., Pylvänäinen, J. W., Rigaud, S. U., et al (2022). TrackMate 7: integrating state-of-the-art segmentation algorithms into tracking pipelines. Nature Methods, 19(7), 829–832. doi:10.1038/s41592-022-01507-1

2. Smolander J., et al (2021). ILoReg: a tool for high-resolution cell population identification from single-cell RNA-seq data. Bioinformatics, 37(8):1107-1114. doi: 10.1093/bioinformatics/btaa919.

